pkg/virtual/workspaces/registry/rest.go

```diff
@@ -0,0 +1,482 @@
+func NewREST(workspaceClient workspaceClient.WorkspaceInterface, rbacClient rbacv1client.RbacV1Interface, crbInformer rbacinformers.ClusterRoleBindingInformer, workspaceReviewerProvider workspaceauth.ReviewerProvider, workspaceLister workspaceauth.Lister) *REST {
+	return &REST{
+		rbacClient:                rbacClient,
+		crbInformer:               crbInformer,
+		crbLister:                 crbInformer.Lister(),
+		workspaceReviewerProvider: workspaceReviewerProvider,
+		workspaceClient:           workspaceClient,
+		workspaceLister:           workspaceLister,
+		createStrategy:            Strategy,
+		updateStrategy:            Strategy,
+
+		TableConvertor: printerstorage.TableConvertor{TableGenerator: printers.NewTableGenerator().With(workspaceprinters.AddWorkspacePrintHandlers)},
+	}
+}
+
+// New returns a new Workspace
+func (s *REST) New() runtime.Object {
+	return &tenancyv1alpha1.Workspace{}
+}
+
+// NewList returns a new WorkspaceList
+func (*REST) NewList() runtime.Object {
+	return &tenancyv1alpha1.WorkspaceList{}
+}
+
+func (s *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (s *REST) getPrettyNameFromInternalName(user kuser.Info, internalName string) (string, error) {
+	list, err := s.crbInformer.Informer().GetIndexer().ByIndex(InternalNameIndex, internalName)
+	if err != nil {
+		return "", err
+	}
+	for _, el := range list {
+		if crb, isCRB := el.(*rbacv1.ClusterRoleBinding); isCRB &&
+			len(crb.Subjects) == 1 && crb.Subjects[0].Name == user.GetName() {
+			return crb.Labels[PrettyNameLabel], nil
+		}
+	}
+	return "", kerrors.NewNotFound(tenancyv1alpha1.SchemeGroupVersion.WithResource("workspaces").GroupResource(), internalName)
+}
+
+func (s *REST) getInternalNameFromPrettyName(user kuser.Info, prettyName string) (string, error) {
+	list, err := s.crbInformer.Informer().GetIndexer().ByIndex(PrettyNameIndex, prettyName)
+	if err != nil {
+		return "", err
+	}
+	for _, el := range list {
+		if crb, isCRB := el.(*rbacv1.ClusterRoleBinding); isCRB &&
+			len(crb.Subjects) == 1 && crb.Subjects[0].Name == user.GetName() {
+			return crb.Labels[InternalNameLabel], nil
+		}
+	}
+	return "", kerrors.NewNotFound(tenancyv1alpha1.SchemeGroupVersion.WithResource("workspaces").GroupResource(), prettyName)
+}
+
+// List retrieves a list of Workspaces that match label.
+func (s *REST) List(ctx context.Context, options *metainternal.ListOptions) (runtime.Object, error) {
+	user, ok := apirequest.UserFrom(ctx)
+	if !ok {
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspace"), "", fmt.Errorf("unable to list workspaces without a user on the context"))
+	}
+
+	if scope := ctx.Value(WorkspacesScopeKey); scope == personalScope {
+		user = &kuser.DefaultInfo{
+			Name:   user.GetName(),
+			UID:    user.GetUID(),
+			Groups: []string{},
+			Extra:  user.GetExtra(),
+		}
+	}
+
+	labelSelector, _ := InternalListOptionsToSelectors(options)
+	workspaceList, err := s.workspaceLister.List(user, labelSelector)
```

Comment:
This lister is informer driven, isn't it? Here we have a big (and important) TODO: the lister can be stale. It breaks the API guarantees of lists. To make it correct we have to know the latest RV of the org workspace shard, and then wait for freshness relative to that RV of the lister. This needs a giant TODO here and an issue in the github project.

Reply:
TODO added